### PR TITLE
Simplify the validation in Run call

### PR DESCRIPTION
**Pull request by linkerzhang**
This PR contains,
1. Simplify the validation for inputs/outputs.
     Whether required inputs are fed should/will be detected when running (as there's no data, nullptr should be there).
2. Remove the lock in Run call to check whether session is initialized, which is unnecessary.

### CODE REVIEW

File: onnxruntime/core/session/inference_session.cc [hunk 1]
```diff
@@ -618,12 +571,9 @@ Status InferenceSession::Run(const RunOptions& run_options,
   Status retval = Status::OK();
 
   try {
-    {
-      std::lock_guard<onnxruntime::OrtMutex> l(session_mutex_);
-      if (!is_inited_) {
-        LOGS(*session_logger_, ERROR) << "Session was not initialized";
-        retval = Status(common::ONNXRUNTIME, common::FAIL, "Session not initialized.");
-      }
+    if (!is_inited_) {
```
Comment (pranavsharma) [inline onnxruntime/core/session/inference_session.cc]: is_inited_ should be declared std::atomic<bool> if you want to remove the lock here. In fact the C++ standard has a dedicated type for it called atomic_bool. https://en.cppreference.com/w/cpp/atomic/atomic

Comment (linkerzhang) [reply onnxruntime/core/session/inference_session.cc]: it's only needed when writing access to is_inited_, I think.

Comment (linkerzhang) [reply onnxruntime/core/session/inference_session.cc]: normally, callers should, 1. create/initialize session. 2) multi-threading call run if he/she's willing to. In that case, it's not needed to have a lock here. in worst case, callers call initialize in separate thread, no lock here will only fail one run call? btw, the lock is there when editing is_inited_, I believe.

Comment (yuanbyu) [reply onnxruntime/core/session/inference_session.cc]: The safest is to make is_init_ atomic. But I am inclined to believe we are ok here. In Ke's case of calling init in a thread parallel to run call, it is a bug regardless of whatever locks you use.

**Pull request merged**